core/src/services/dbfs/reader.rs

```diff
@@ -0,0 +1,183 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use std::cmp;
+use std::io::SeekFrom;
+use std::sync::Arc;
+use std::task::ready;
+use std::task::Context;
+use std::task::Poll;
+
+use async_trait::async_trait;
+use base64::engine::general_purpose;
+use base64::Engine;
+use bytes::BufMut;
+use bytes::Bytes;
+use bytes::BytesMut;
+use futures::future::BoxFuture;
+use http::Response;
+use serde::Deserialize;
+
+use super::core::DbfsCore;
+
+use crate::raw::oio::WriteBuf;
+use crate::raw::*;
+use crate::*;
+
+const DBFS_READ_LIMIT: usize = 1048576;
+
+pub struct DbfsReader {
+    state: State,
+    path: String,
+    offset: u64,
+    buffer: BytesMut,
+}
+
+impl DbfsReader {
+    pub fn new(core: Arc<DbfsCore>, op: OpRead, path: String, content_length: u64) -> Self {
+        DbfsReader {
+            state: State::Idle(Some(core)),
+            path,
+            offset: op.range().offset().unwrap_or(0),
+            buffer: BytesMut::with_capacity(content_length as usize),
+        }
+    }
+
+    #[inline]
+    fn set_offset(&mut self, offset: u64) {
+        self.offset = offset;
+    }
+
+    fn serde_json_decode(&self, bs: &Bytes) -> Result<Bytes> {
+        let mut response_body = match serde_json::from_slice::<ReadContentJsonResponse>(bs) {
+            Ok(v) => v,
+            Err(err) => {
+                return Err(
+                    Error::new(ErrorKind::Unexpected, "parse response content failed")
+                        .with_operation("http_util::IncomingDbfsAsyncBody::poll_read")
+                        .set_source(err),
+                );
+            }
+        };
+
+        response_body.data = general_purpose::STANDARD
```

Comment:
It doesn't make sense to change the `response_body.data` and it's also wrong that assume the returning data is valid UTF-8. We should return the data decoded directly.

Reply:
Done